### ISSUE TITLE
question about affine optimization for frotran program when using flang

### ISSUE THREAD

**Issue by qelk123** [llvm/llvm-project#58646]
Hi all,
        I am trying to flang-affine-promotion pass to convert parts of my fir code to affine dialect.However,it failed when converting fir.do_loop because of error in fetching loop analysis in AffineFunctionAnalysis.I only saw a case using fir in the repo.Are here any cases writing in fortran that I can refer to?
Thanks

**Comment by kiranchandramohan** [comment]
I think the affine dialect conversion in flang has not been maintained. To get it to work you might need a sequence of passes to run before affine conversion. The `fir-dev` branch of `f18-llvm-project` has an example https://github.com/flang-compiler/f18-llvm-project/blob/fir-dev/flang/test/Fir/arr-end-end.f90. Not sure whether all the passes that are required are present in `llvm/flang` and whether that example will work in `llvm/flang`.

**Comment by qelk123** [comment]
> I think the affine dialect conversion in flang has not been maintained. To get it to work you might need a sequence of passes to run before affine conversion. The `fir-dev` branch of `f18-llvm-project` has an example https://github.com/flang-compiler/f18-llvm-project/blob/fir-dev/flang/test/Fir/arr-end-end.f90. Not sure whether all the passes that are required are present in `llvm/flang` and whether that example will work in `llvm/flang`.
> 
> A two year old presentation by the engineer who worked on this topic is available at https://slides.com/rajanwalia/deck.

Unfortunately,some passes are missing in the `llvm/flang` when I want to reprodece the case.logs as follow:
`tco: Unknown command line argument '--fir-loop-result-opt'.  Try: 'tco --help'
tco: Did you mean '--affine-loop-unroll'?
tco: Unknown command line argument '--simplify-affine-structures'.  Try: 'tco --help'
tco: Did you mean '--simplify-intrinsics'?
tco: Unknown command line argument '--memref-dataflow-opt'.  Try: 'tco --help'
tco: Did you mean '--fir-memref-dataflow-opt'?`
when I simply delete these passes, I still can't promote fir due to the error
`AffineLoopAnalysis: array coordinate is not a loop induction variable (not a block argument)`

**Comment by kiranchandramohan** [comment]
I don't think there is a substitute for `fir-loop-result-opt`. If you are interested and have relevant experience, you can try to upstream that pass from `fir-dev` of `f18-llvm-project` with suitable modifications.

**Comment by qelk123** [comment]
> I don't think there is a substitute for `fir-loop-result-opt`. If you are interested and have relevant experience, you can try to upstream that pass from `fir-dev` of `f18-llvm-project` with suitable modifications.

Thank you,I'll have a try.And there is another question,can I suppose `fir-dev` of `f18-llvm-project` is better developed even though the last commit was on  `22 Apr`?

**Comment by kiranchandramohan** [comment]
No, `llvm/flang` is the most developed version. `fir-dev` of `f18-llvm-project` has not been maintained for a few months now. It might have a few things that we chose not to upstream due to various reasons.

**Comment by kiranchandramohan** [comment]
There might have been some changes required so that loop optimizations in llvm work properly. Could you check whether https://reviews.llvm.org/D132176 introduced `%arg4`?

**Comment by cathieO** [comment]
@qelk123 did you find a solution for this? I have just started looking into affine transformations for fortran codes and it looks like this change enables some llvm loop optimizations and breaks mlir loop optimizations. Have you found a workaround?
